### PR TITLE
Reduction zero dimension extent fixes

### PR DESCRIPTION
**Pull request by SimonHeybrock**
- Fix reduction operations for cases where reduction dim or other dim have extent 0.
- Fix missing unit in `sum` with `out` arg.
- Unrelated: Use correct exception type.

### CODE REVIEW

File: variable/test/creation_test.cpp [hunk 1]
```diff
@@ -44,3 +44,60 @@ TEST_P(DenseVariablesTest, empty_like) {
...
+            makeVariable<int64_t>(var.dims(), var.unit(), Values{0, 0}));
+  EXPECT_EQ(special_like(var, variable::FillValue::Max),
+            makeVariable<bool>(var.dims(), var.unit(),
+                               Values{std::numeric_limits<bool>::max(),
+                                      std::numeric_limits<bool>::max()}));
+  EXPECT_EQ(special_like(var, variable::FillValue::Lowest),
+            makeVariable<bool>(var.dims(), var.unit(),
+                               Values{std::numeric_limits<bool>::lowest(),
+                                      std::numeric_limits<bool>::lowest()}));
+}
```
Comment (OwenArnold) [inline variable/test/creation_test.cpp:103]: I think you could reduce the amount of code here with TYPED_TEST_SUITE

File: variable/reduction.cpp [hunk 2]
```diff
@@ -205,15 +218,17 @@ void max_impl(const VariableView &out, const VariableConstView &var) {
...
 }
 
 /// Return the maximum along given dimension ignoring NaN values.
 ///
 /// Variances are not considered when determining the maximum. If present, the
 /// variance of the maximum element is returned.
 Variable nanmax(const VariableConstView &var, const Dim dim) {
-  return reduce_idempotent(var, dim, core::element::nanmax_equals);
+  return reduce_idempotent(var, dim, core::element::nanmax_equals,
+                           FillValue::Lowest);
```
Comment (OwenArnold) [inline variable/reduction.cpp:231]: `FillValue::Lowest` used here, but equivalent for `min` below i` `FillValue::Max` -> should this be `FillValue::Min` rather than`FillValue::Lowest`

File: variable/reduction.cpp [hunk 3]
```diff
@@ -38,13 +51,9 @@ void nansum_impl(const VariableView &summed, const VariableConstView &var) {
...
 template <typename Op>
 Variable sum_with_dim_impl(Op op, const VariableConstView &var, const Dim dim) {
-  auto dims = var.dims();
-  dims.erase(dim);
   // Bool DType is a bit special in that it cannot contain it's sum.
   // Instead the sum is stored in a int64_t Variable
-  Variable summed{is_dtype_bool(var) ? makeVariable<int64_t>(Dimensions(dims))
-                                     : copy(var.slice({dim, 0}))};
-  fill_zeros(summed);
+  auto summed = make_accumulant(var, dim, FillValue::ZeroNotBool);
```
Comment (OwenArnold) [inline variable/reduction.cpp:56]: I'm unclear from the PR description where the original bug was found. Is this were there error was in the slice with   0-D data?

File: variable/test/sum_test.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+TEST_F(SumTest, sum) {
+  const auto expectedX =
+      makeVariable<double>(Dims{Dim::Y}, Shape{2}, units::m, Values{3.0, 7.0});
+  const auto expectedY =
+      makeVariable<double>(Dims{Dim::X}, Shape{2}, units::m, Values{4.0, 6.0});
+  EXPECT_EQ(sum(var, Dim::X), expectedX);
+  EXPECT_EQ(sum(var, Dim::Y), expectedY);
+}
+
+TEST_F(SumTest, sum_with_empty_dim) {
```
Comment (OwenArnold) [inline variable/test/sum_test.cpp:30]: 👍

**Review by OwenArnold — CHANGES_REQUESTED** [review]
Good improvements including exception types and test refactoring in the process.

File: variable/test/creation_test.cpp (hunk 1 above)
Comment (SimonHeybrock) [reply variable/test/creation_test.cpp:103]: I don't think so, there are too many differences if you look closely.

Comment (jl-wynen) [inline variable/test/creation_test.cpp:103]: - Given that we found theses errors because of problems with empty variables, it would make sense to also test these with empty prototypes.
- What about binned variables?

File: variable/reduction.cpp (hunk 2 above)
Comment (SimonHeybrock) [reply variable/reduction.cpp:231]: Not sure I understand your question. `Min` would be close to 0, not `-max`.

File: variable/reduction.cpp (hunk 3 above)
Comment (SimonHeybrock) [reply variable/reduction.cpp:56]: Yes, `slice({dim, 0})` was an out-of-bounds access.

File: variable/test/reduce_various_test.cpp [hunk 5]
```diff
@@ -47,6 +47,22 @@ TEST(ReduceTest, min_max_with_variances) {
...
 
+TEST(ReduceTest, min_max_empty_dim) {
+  const auto var = makeVariable<double>(Dims{Dim::X, Dim::Y}, Shape{2, 0},
+                                        Values{}, Variances{});
+  EXPECT_EQ(max(var, Dim::X), makeVariable<double>(Dims{Dim::Y}, Shape{0},
+                                                   Values{}, Variances{}));
+  const auto highest = std::numeric_limits<double>::max();
+  EXPECT_EQ(max(var, Dim::Y),
+            makeVariable<double>(Dims{Dim::X}, Shape{2},
+                                 Values{-highest, -highest}, Variances{0, 0}));
```
Comment (jl-wynen) [inline variable/test/reduce_various_test.cpp:58]: Why did you use `-highest` instead of `lowest`?

File: variable/reduction.cpp (hunk 2 above)
Comment (OwenArnold) [reply variable/reduction.cpp:231]: Indeed. Lowest is the correct solution.

**Review by OwenArnold — APPROVED** [review]

File: core/include/scipp/core/element/creation.h [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+template <class T> struct underlying { using type = T; };
+template <class T> struct underlying<ValueAndVariance<T>> { using type = T; };
+template <class T> using underlying_t = typename underlying<T>::type;
+
+constexpr auto numeric_limits_max_like =
+    overloaded{special_like, [](const auto &x) {
+                 return std::numeric_limits<
+                     underlying_t<std::decay_t<decltype(x)>>>::max();
+               }};
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply core/include/scipp/core/element/creation.h:41]: Yes, so `transform` determines whether the output has variances based on whether any of the input has variances. This will then assign the value, which results in variances getting set to zero.

For this application we do not need more control. For others you can make a kernel that returns value and variance? Not sure this is relevant here?

File: variable/test/reduce_various_test.cpp (hunk 5 above)
Comment (SimonHeybrock) [reply variable/test/reduce_various_test.cpp:58]: To ensure I did not mistunderstand `min` vs. `lowest`.

File: core/include/scipp/core/element/creation.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (jl-wynen) [reply core/include/scipp/core/element/creation.h:41]: Probably not relevant. I was just wandering if this is consistent with other kernels.

File: variable/test/creation_test.cpp (hunk 1 above)
Comment (SimonHeybrock) [reply variable/test/creation_test.cpp:103]: >     * Given that we found theses errors because of problems with empty variables, it would make sense to also test these with empty prototypes.

I think this is a problem accross the codebase, I suggest addressing this as a larger maintenance task.

Comment (jl-wynen) [reply variable/test/creation_test.cpp:103]: Ok, did you run some simple manual tests to make sure it's not completely broken?

Comment (SimonHeybrock) [reply variable/test/creation_test.cpp:103]: >     * What about binned variables?

This is actually a generic shortcoming(?) of the test for many of our operations. I think the answer is:
- Since it is just using `transform` with nothing else around, it should not be necessary to add such test for every function --- provided the `transform` tests are good, we do not need to test for every `dtype`, just like we don't generally test every function with a complete list of dtypes.
- We should extend what I have started with `DenseVariablesTest` (in `variable/test/test_variables.h`) to be used in more places, and also adding `BinnedVariablesTest`. Again, this is a larger task and we should schedule it as maintenance.

File: variable/test/reduce_various_test.cpp (hunk 5 above)
Comment (jl-wynen) [reply variable/test/reduce_various_test.cpp:58]: Hm, doesn't feel like this should be part of this test

File: variable/test/creation_test.cpp (hunk 1 above)
Comment (SimonHeybrock) [reply variable/test/creation_test.cpp:103]: > Ok, did you run some simple manual tests to make sure it's not completely broken?

No, because I *know* it is completely broken. Most functions on the `Variable` API assume it is non-null, i.e., require the API user to ensure it is valid. Note that creating null variables on the Python side is not really possible with the (public) API.

Comment (jl-wynen) [reply variable/test/creation_test.cpp:103]: > Note that creating null variables on the Python side is not really possible with the (public) API.

Unless as the result of some operations.

But then I guess, we will have to wait until we do the bit testing update to find and eliminate those cases.

**Review by jl-wynen — APPROVED** [review]

File: variable/test/reduce_various_test.cpp (hunk 5 above)
Comment (SimonHeybrock) [reply variable/test/reduce_various_test.cpp:58]: https://en.cppreference.com/w/cpp/types/numeric_limits/lowest says it is guarenteed for std floating point types.
